File: data structure/quicksort.cpp
#include <iostream>
using namespace std;
int partition(int arr[] , int low , int high){
    int i, j , temp;
    i=low+1;
    j=high;
    int piviot=arr[low];
    do{
        while (arr[i]<=piviot){
            i++;

        }
        while (arr[j]>piviot){

            j--;
        }
        if (i<j){
            int temp=arr[i];
            arr[i]=arr[j];
            arr[j]=temp;
        }
    }while (i<j);
temp=arr[low];
arr[low]=arr[j];
arr[j]=temp;
return j;
}
void quicksort(int arr[] , int low ,int high){
 if (low < high){
    int piviot=partition(arr, low , high);
    quicksort(arr, low , piviot-1);
    quicksort(arr, piviot+1 , high);

}
}
int main()
{
    // int arr[]={19,9,14,24,29,27};
    int arr[]={1,2,3,4,5,6};       //{24,27,25}
    int n=6;
   quicksort(arr , 0 , n-1);
   for (int  i = 0; i < n; i++)
   {
     cout << arr[i]<< " "; 
   }
    
return 0;
}

File: JAVA/Searching, sorting ,Array and LinkedList/QuickSort.java
public class QuickSort {

    // Function to partition the array and return the pivot index
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high]; // Choose the last element as the pivot
        int i = (low - 1); // Index of smaller element
        
        for (int j = low; j < high; j++) {
            // If the current element is smaller than or equal to the pivot
            if (arr[j] <= pivot) {
                i++;
                
                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i + 1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    // Function to implement QuickSort recursively
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // Partition the array and get the pivot index
            int pivotIndex = partition(arr, low, high);

            // Recursively sort the elements before and after partition
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }

    // Main function to test the QuickSort implementation
    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        int n = arr.length;

        System.out.println("Unsorted array:");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();

        // Call the quickSort function
        quickSort(arr, 0, n - 1);

        System.out.println("Sorted array:");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}
